content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java

```diff
@@ -10384,5 +10412,362 @@ public void doViewTrash(RunData data)
+
+		return TEMPLATE_ZIPDOWNLOAD_FINISH;
+	}
+
+	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException
+	{
+		String action = request.getParameter("eventSubmit_doFinalizeZipDownload");
+
+		if ((action==null)||(action.isEmpty()))
+		{
+			super.doPost(request,response);
+			return;
+		}
+
+		checkRunData(request);
+		JetspeedRunData rundata = (JetspeedRunData) request.getAttribute(ATTR_RUNDATA);
+		SessionState state = rundata.getPortletSessionState (rundata.getJs_peid ());
+		List<ListItem> zipDownloadItems = (List<ListItem>) state.getAttribute(STATE_ZIPDOWNLOAD_SET);
+
+		String collectionId = (String) request.getParameter("collectionId");
+		ZipOutputStream zipOut = null;
+		try
+		{
+			ContentCollection collection = ContentHostingService.getCollection(collectionId);
+			ResourceProperties props = collection.getProperties();
+			String rootFolderName = escapeInvalidCharsEntry(props.getPropertyFormatted(props.getNamePropDisplayName()));
+
+			response.setContentType("application/zip;charset=UTF-8");
+			response.setHeader("Content-Disposition", "attachment; filename = site_download.zip");
+			zipOut = new ZipOutputStream(response.getOutputStream());
+
+			Iterator it = zipDownloadItems.iterator();
+			while(it.hasNext())
+			{
+				ListItem myElement = (ListItem) it.next();
+				String resourceId = myElement.getId();
+				boolean get = ContentHostingService.allowGetResource(resourceId);
+				if (get) compressResource(zipOut, collectionId, rootFolderName, resourceId);
+			}
+		}
+		catch (PermissionException pe)
+		{
+			try
+			{
+				response.sendError(response.SC_FORBIDDEN);
+			}
+			catch (IOException e)
+			{
+				e.printStackTrace();
+			}
+		}
+		catch (Throwable ignore)
+		{
+			try
+			{
+				response.sendError(response.SC_NO_CONTENT);
+			}
+			catch (IOException e)
+			{
+				e.printStackTrace();
+			}
+		}
+		finally
+		{
+			if (zipOut != null)
+			{
+				try
+				{
+					zipOut.flush();
+					zipOut.close();
+				}
+				catch (Throwable ignore)
+				{
+					ignore.printStackTrace();
+				}
+			}
+		}
+	}
+
+	protected void compressResource(ZipOutputStream zipOut, String collectionId, String rootFolderName, String resourceId) throws Exception
+	{
+		if (ContentHostingService.isCollection(resourceId))
+		{
+			try
+			{
+				ContentCollection collection = ContentHostingService.getCollection(resourceId);
+				List<String> children = collection.getMembers();
+				if(children != null)
+				{
+					for(int i = children.size() - 1; i >= 0; i--)
+					{
+						String child = children.get(i);
+						compressResource(zipOut,collectionId,rootFolderName,child);
+					}
+				}
+			}
+			catch (PermissionException e)
+			{
+				//Ignore
+			}
+		}
+		else
+		{
+			try
+			{
+				ContentResource resource = ContentHostingService.getResource(resourceId);
+				ResourceProperties props = resource.getProperties();
+				String displayName = props.getPropertyFormatted(props.getNamePropDisplayName());
```

Comment:
Do you mean that it will fail when files are renamed after uploading them? If not, could you give me a little test plan to check if it fails with the current patch and fix it?